Symex-dereference: simplify after deref [blocks: #2574, #4056]

tautschnig opened the pull request:
This is generally useful because it turns member-of-if into if-of-member, which
avoids repeatedly quoting large complex if expressions in a with_exprt, but it
is especially good for field sensitivity as it puts member expressions directly
on top of symbol expressions, which field sensitivity knows how to deal with.

Performance evaluation is required ~~, but also this change shows that some tests newly fail with this change, which is unexpected. See #3724 for the first one of them~~.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

smowton committed:
Symex-dereference: simplify after deref

This is generally useful because it turns member-of-if into if-of-member, which
avoids repeatedly quoting large complex if expressions in a with_exprt, but it
is especially good for field sensitivity as it puts member expressions directly
on top of symbol expressions, which field sensitivity knows how to deal with.

tautschnig commented:
This requires the byte_update fixes from #2068 to make the SMT solver happy (adding the simplify step changes a nested `byte_extract` into a `byte_update`).

allredj commented on src/goto-symex/symex_dereference.cpp:
```diff
@@ -389,4 +390,23 @@ void goto_symext::dereference(
+  // when all we need is
+  // s1=s1 with member:=X [and guard b]
+  // s2=s2 with member:=X [and guard !b]
+  do_simplify(expr);
```
I'm curious to see what the impact on performance this change has.

tautschnig replied:
I will benchmark (and I hope @romainbrenguier can do his benchmarking as well) once the dependencies #2068 and #3770 are merged so that we have a precise analysis.

romainbrenguier replied:
This could maybe be made faster with a simplify method which would focus on the conversions that are mentioned above.

tautschnig replied:
That is certainly one option - are there particular patterns that affect Java? (@smowton maybe?)

kroening replied:
Update: this appears to be causing performance degradation; we may need to look into a specialised simplifier to achieve this.

allredj commented on src/goto-symex/symex_dereference.cpp:
```diff
@@ -389,4 +390,23 @@ void goto_symext::dereference(
+  // make sure simplify has not re-introduced any dereferencing that
+  // had previously been cleaned away
+  INVARIANT(
+    !has_subexpr(expr, ID_dereference), "simplify re-introduced dereferencing");
```
This seems quite heavy to run on each call to `dereference`. Do we yet have a way to avoid evaluating invariants (and other assertions) in deployed binaries of CBMC?

tautschnig replied:
Good point, I have now wrapped it in `if(symex_config.run_validation_checks)`.

allredj commented on src/goto-symex/symex_dereference.cpp:
```diff
@@ -389,4 +390,23 @@ void goto_symext::dereference(
+  // applying field sensitivity, which can turn such field-of-symbol expressions
+  // into atomic SSA expressions, but would have to rewrite all of 'o1'
+  // otherwise.
+  // make the structure of the lhs as simple as possible to avoid,
```
This line seems redundant

tautschnig replied:
Comment cleaned up.

smowton commented:
Preliminary benchmarks results (note small numbers, running a bigger benchmark now):
![perf_out](https://user-images.githubusercontent.com/283483/53256480-e4fdc300-36bf-11e9-9476-db287d0a7e3d.png)

tautschnig commented:
@smowton Just checking: is this current develop (lacking #2068, #3770) against #2068+#3770+this PR?

smowton commented:
No #3770, otherwise yes that's accurate. I asked @romainbrenguier for advice to produce a more "substantial" benchmark, but currently I get a highly bimodal distribution (lots of very fast results and a small number of very slow ones) on both develop and on this branch, which is hardly ideal. I'll try harder on Monday.

tautschnig commented:
I'm not sure whether benchmarks completing in less than one second have a good signal to noise ratio. You'd need a very deterministic and clean environment to get solid data out of this when all that you are measuring is total runtime.

tautschnig commented:
I have started a benchmarking run, should be able to provide results this evening. I would, however, also encourage benchmarking on Java code as there may be considerable differences.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 12aa0fc4492ce66cd31dd63bd184db2eaf8d6ff6).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/102519719
Status will be re-evaluated on next push.
Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

  - the compatibility was already broken by an earlier merge.

smowton commented:
Time to run Symex on a few Tika methods:

![perf_out](https://user-images.githubusercontent.com/283483/53505944-d55df000-3aac-11e9-914e-fe71fafd7c25.png)

(so no particular difference seen here)

tautschnig commented:
I will provide my bit of data as soon as available, but @kroening @peterschrammel this is ready for review. (Most of the +39 lines added in this PR are comments...)

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: e151eeeb459952a974484cb5a27d0baf7dab5f7b).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/102538323

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: aeda45a89ad7ef91ad4e6969ff3f6780e8004ae4).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/102572084

tautschnig commented:
Data finally available: there is a slowdown (as expected for my benchmark set), but nothing that I find particularly scary. On SV-COMP's ReachSafety-ECA (which basically does no pointer operations at all) the number of `symex_step` invocations per second decreases from 2555.08 down to 2309.25. While this is a 10% slowdown, my branch collecting the various performance fixes did 4978.76 when I last checked on the same benchmarks so I really wouldn't care about the slowdown on some benchmarks if on the other hand this helps Java/code heavily using pointers.

romainbrenguier commented:
@smowton 
> (so no particular difference seen here)

It seems to be slightly slower on average. Can you have give the total time numbers?

allredj commented:
Thanks for running those benchmarks. Is there any evidence that this helps for Java?

smowton commented:
Summary here: this isn't expected to cause a performance increase by itself (and indeed will definitely cost a little more). It is primarily a pre-requisite for field-sensitive symex, because of the member-of-if -> if-of-member transformation mentioned in the comments added by this PR.

I don't see a clear way to restrict simplify's activities except to keep track of expressions which are already maximally simplified and suppress repeated simplification. I've seen a couple of drafts by @tautschnig which attempt this, either by the simplifier tracking ireps it has seen and simplified, or by attaching comments to simplified ireps, or by adding an extra boolean to `irept` to track is-simplified state.

romainbrenguier commented:
> Summary here: this isn't expected to cause a performance increase by itself (and indeed will definitely cost a little more). It is primarily a pre-requisite for field-sensitive symex, because of the member-of-if -> if-of-member transformation mentioned in the comments added by this PR.

I guess that's fine, but in that case it could have been better for it to be part of the field-sensitive PR.

tautschnig commented:
> I've seen a couple of drafts by @tautschnig which attempt this, either by the simplifier tracking ireps it has seen and simplified, or by attaching comments to simplified ireps, or by adding an extra boolean to `irept` to track is-simplified state.

FWIW, https://github.com/tautschnig/cbmc/commit/ef00e0a61b53e93555eb9e0cc37ff153c0bc4458 is the most serious of those. It does seem to work well, but @kroening rightfully remarked that it opens the door to everyone asking for their little pet flag. But we'll get there when the other performance-optimising PRs have been reviewed/merged so that I can start posting those changes as PRs. Yet, as @smowton said, it's one more reason I'm not at all worried about performance cost here.

I have no issue merging this as part of #2574 as @romainbrenguier suggested, but I think it was important to debug and evaluate it separately as this has lead to a lot of type-related bug fixes.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 9abfc972c8e162e578bea47019b28bfa03167149).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/102907891

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: f1ce452113207f3217f79fcd4a27f8696e886a3c).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/103024631
Status will be re-evaluated on next push.
Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

  - the compatibility was already broken by an earlier merge.

tautschnig commented:
@smowton will add some benchmarking data to confirm that this is necessary for #4314 to go ahead.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 427df5c06114065b94b7747227c58ba22a14075f).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/103062461
Status will be re-evaluated on next push.
Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

  - the compatibility was already broken by an earlier merge.

smowton commented:
I benchmarked 133 Tika methods and found total runtime = 4818s with this PR + #4314, 4860s with just #4314, 5974s with just develop. Most of that reduction comes from 3 tests that were previously timeouts (timeout = 300s) that now complete in 30 seconds; the rest comes from 10-15s reductions in many tests that previously took 10-20s to analyse.

I expected to find a major penalty for not including this PR alongside #4314, because #4314 causes `value_set_dereferencet` to more rigorously use `byte_extract` operations when generic types differ, but no such effect was observed. This may be because constant-propagation was rarely succeeding in these cases and without it the byte-extract op is not actually that expensive, or it may be because the associated cost is only exposed when field-sensitivity is enabled, which by failing to promote member-of-symbol to an atomic SSA identifier forces a more-expensive composite struct operation.

smowton commented:
4314 + this + develop vs. develop: 
![perf_out](https://user-images.githubusercontent.com/283483/53882595-a7c5f900-400e-11e9-87f6-ef0a5cb5cf97.png)

smowton commented:
Successfully reproduced a micro-benchmark exposing the kind of problem I was expecting:

Test program:
```java
public class Test {

  public static void main() {

    Generic<Integer> g = new GenericSub<Integer>();

    int x = 0;
    for(int i = 0; i < 3000; ++i)
      x += g.get();

    assert x == 0;

  }

}

class Generic<T> {

  T key; // Only here to make sure the compiler doesn't throw <T> away
  int x;

  public int get() { return 0; }

  public Generic() {
    key = null;
    x = 5;
  }

}

class GenericSub<S> extends Generic<S> {

  public int get() { return x; }

}
```

Times for `--show-vcc`:

Develop: 1.3s
#4314: 9.4s
#4314 + #3725: 1.5s

smowton commented:
Excerpt from `--show-vcc` illustrating the problem:

```
-{-14} symex_dynamic::dynamic_object1#3 = { { { "java::GenericSub" }, NULL, 0 } }
-{-15} symex_dynamic::dynamic_object1#4 = { { { "java::GenericSub" }, NULL, 5 } }
+{-14} symex_dynamic::dynamic_object1#3 = byte_extract_little_endian(with(byte_extract_little_endian({ { "java::GenericSub" }, NULL, 0 }, 0, struct java::Generic), member_name component_name="key", NULL), 0, struct java::GenericSub)
+{-15} symex_dynamic::dynamic_object1#4 = byte_extract_little_endian(with(byte_extract_little_endian(symex_dynamic::dynamic_object1#3.@Generic, 0, struct java::Generic), member_name component_name="x", 5), 0, struct java::GenericSub)
```

smowton commented:
VCCs for develop and #4314 + #3725 are identical; those for #4314 alone are ~25000 lines longer due to constant prop failure

tautschnig committed:
Test to make sure generics do not result in constant propagation failure

This was fine with develop with and without simplification after dereferencing,
but would break if removing base_type_eq without adding the simplification step.

tautschnig commented:
Thanks a lot @smowton for the detailed analysis! I have added the test that you provided as a second commit.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 0b09ffa61bae058358f94f5e1beaeb5e3b53ece5).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/103459224
